pennylane/_grad.py

```diff
@@ -93,16 +94,25 @@ def _get_grad_fn(self, args):
         if len(argnum) == 1:
             argnum = argnum[0]
 
-        return self._grad_with_forward(
-            self._fun,
-            argnum=argnum,
-        )
+        return self._grad_with_forward(self._fun, argnum=argnum), argnum
 
     def __call__(self, *args, **kwargs):
         """Evaluates the gradient function, and saves the function value
         calculated during the forward pass in :attr:`.forward`."""
-        grad_value, ans = self._get_grad_fn(args)(*args, **kwargs)
+        grad_fn, argnum = self._get_grad_fn(args)
+
+        if not isinstance(argnum, int) and not argnum:
+            warnings.warn(
+                "Attempted to differentiate a function with no trainable parameters. "
+                "If this is unintended, please add trainable parameters via the "
+                "`requires_grad` attribute or `argnum` keyword."
+            )
+            self._forward = self._fun(*args, **kwargs)
+            return ()
```

Comment:
I thought it might be good to bypass invoking the grad function in this case, but let me know if this is correct (and hopefully doesn't cause any issues)>

Reply:
The computation is likely very similar, but this has the advantage of not generating the autograd warning.

Reply:
🤔 I can't think of any edge cases, but I _might_ lean to simply allowing the grad function to be called simply to avoid any unintentional bugs/edge cases?

Reply:
I would be inclined to use this, unless there is strong opposition. That way the autograd warning is suppressed, and we clearly define what the return value is for a gradient with no trainable parameters. The jacobian behaves the same way by returning an empty tuple without ever invoking autograd when there are no trainable parameters.

Reply:
Hi @quantshah, thank you for providing feedback!

The discussion here was in regards to suppressing the warning generated by Autograd, because the warning message is not necessarily clear and we added a better warning on the PennyLane side.

Providing a regular list as a QNode parameter should not compute the gradient, as QNode parameters are no longer assumed trainable with the new release v0.21.0. To compute the gradient of a list parameter you have two options:
- wrap the list in a PennyLane NumPy array (as you already do), and the trainability can also be explicitly set: `qml.grad(circuit)(np.array([0.1, 0.2, 0.3], requires_grad=True))`
- use the `argnum` argument: `qml.grad(circuit, argnum=0)([0.1, 0.2, 0.3])`

Unfortunately, I cannot reproduce your output on either the latest master branch or the most recent release. The output I get includes the warning for empty gradients:
```py
-0.09983341664682799
[-9.95004165e-01 -1.21430643e-17 -1.21430643e-17]
()
/home/dave/.local/lib/python3.8/site-packages/pennylane/_grad.py:105: UserWarning: Attempted to differentiate a function with no trainable parameters. If this is unintended, please add trainable parameters via the 'requires_grad' attribute or 'argnum' keyword.
  warnings.warn(
```

If the missing warning persists for you on the latest release, it would be great if you could open an issue with your system information and software versions.

Reply:
Thanks @dime10. I see now that the warning message actually tells us that we should not call gradients in this way. Feel free to ignore my comments then. But I could not figure out the issue before just from the warning message and realised it after reading the discussion on this PR. I thought the warning was about not having any parameters as inputs to the gradient function (which I was providing as a list and it worked find for the circuit). 

Just a suggestion for some more details in the warning: "Invalid trainable parameters: Inputs to grad function should be valid trainable parameters constructed using one of the supported interfaces (np, jax, tf, pytorch). If this is unintended, please add trainable parameters via the 'requires_grad' attribute or 'argnum' keyword. ”.